### ISSUE TITLE
net/textproto: reduce allocations in ReadMIMEHeader

### ISSUE THREAD

**Issue by rs** [golang/go#37809]
### What version of Go are you using (`go version`)?

<pre>
1.14
</pre>

### What did you do?

For our use-case (a HTTP proxy) `textproto.ReadMIMEHeader` is in the top 3 in term of allocations. In order to ease the GC, we could pre-allocate 512 bytes to be used for small header values. This can reduce by half the number of allocations for this function.

```
name                              old time/op    new time/op    delta
ReadMIMEHeader/client_headers-16    2.30µs ± 5%    2.11µs ± 2%   -8.21%  (p=0.000 n=10+10)
ReadMIMEHeader/server_headers-16    1.94µs ± 3%    1.85µs ± 3%   -4.56%  (p=0.000 n=10+10)

name                              old alloc/op   new alloc/op   delta
ReadMIMEHeader/client_headers-16    1.53kB ± 0%    1.69kB ± 0%  +10.48%  (p=0.000 n=10+10)
ReadMIMEHeader/server_headers-16    1.09kB ± 0%    1.44kB ± 0%  +32.32%  (p=0.000 n=10+10)

name                              old allocs/op  new allocs/op  delta
ReadMIMEHeader/client_headers-16      14.0 ± 0%       6.0 ± 0%  -57.14%  (p=0.000 n=10+10)
ReadMIMEHeader/server_headers-16      14.0 ± 0%       6.0 ± 0%  -57.14%  (p=0.000 n=10+10)
```

The patch is pretty small:

```
index d26e981ae4..6126a6685c 100644
--- a/src/net/textproto/reader.go
+++ b/src/net/textproto/reader.go
@@ -13,6 +13,7 @@ import (
        "strconv"
        "strings"
        "sync"
+       "unsafe"
 )

 // A Reader implements convenience methods for reading requests
@@ -502,6 +503,12 @@ func (r *Reader) ReadMIMEHeader() (MIMEHeader, error) {
                return m, ProtocolError("malformed MIME header initial line: " + string(line))
        }

+       // Create a pre-allocated byte slice for all header values to save
+       // allocations for the first 512 bytes of values, a size that fit typical
+       // small header values, larger ones will get their own allocation.
+       const valuesPreAllocSize = 1 << 9
+       valuesPreAlloc := make([]byte, 0, valuesPreAllocSize)
+
        for {
                kv, err := r.readContinuedLineSlice(mustHaveFieldNameColon)
                if len(kv) == 0 {
@@ -527,7 +534,17 @@ func (r *Reader) ReadMIMEHeader() (MIMEHeader, error) {
                for i < len(kv) && (kv[i] == ' ' || kv[i] == '\t') {
                        i++
                }
-               value := string(kv[i:])
+
+               // Try to fit the value in the pre-allocated buffer to save allocations.
+               var value string
+               if len(kv[i:]) <= valuesPreAllocSize-len(valuesPreAlloc) {
+                       off := len(valuesPreAlloc)
+                       valuesPreAlloc = append(valuesPreAlloc, kv[i:]...)
+                       v := valuesPreAlloc[off:]
+                       value = *(*string)(unsafe.Pointer(&v))
+               } else {
+                       value = string(kv[i:])
+               }

                vv := m[key]
                if vv == nil && len(strs) > 0 {
```

Please tell me if it's worth submitting a PR with this change.

**Comment by smasher164** [comment]
I wonder if it's possible to make use of `strings.Builder` and pre-allocating with `Grow(valuesPreAllocSize)`, instead of depending on unsafe and only using the buffer when the value fits inside it.

**Comment by bradfitz** [comment]
We're not adding unsafe usage to textproto.

**Comment by rs** [comment]
With or without `strings.Builder`, this trick won't work if you grow the slice, as a copy of the backing array will be performed, which would be a waste: the strings created before the grow would still point on the old backing array, copying data for nothing and wasting the first half of the new grown backing array.

I tried a slightly more complex implementation that would cover requests with arbitrary number header values, but it added no gain from my training set of real-life requests/responses.

We could still use `strings.Builder` to hide the use of unsafe, but I'm not sure to see the point.

**Comment by rs** [comment]
> We're not adding unsafe usage to textproto.

@bradfitz would it be accepted if hidden by the use of `strings.Builder`?

**Comment by rs** [comment]
@bradfitz would you be open to adding the `Slice(off, len int) string` to `strings.Builder` to open for implementation of such trick without the need of `unsafe`?

**Comment by bcmills** [comment]
@rs, could you write all of the values into the `strings.Builder` and record the offsets, then convert all of the offsets to slices of the final `string` result?

**Comment by rs** [comment]
That could work yes, but it would make the code slightly more complex. I would personally prefer the `Slice` method on `strings.Builder`, but I guess it will require some more convincing :)

**Comment by josharian** [comment]
Another option (which would also complicate the code) is to have a fixed size, stack-allocated buffer for pending values to record and for the bytes involved. When a new kv pair is available, append the bytes to the bytes buffer and indices in the bytes buffer to the k/v pairs buffer. When either buffer is too full, convert the byte buffer to a string and add slices of that string to the map. This could all be wrapped up methods in a separate type, for sanity. (This is similar to the strings.Builder idea, but without the extra strings.Builder layer, since a plain ol' byte slice should suffice.)

Also, if the keys are commonly re-used, string interning might help.

**Comment by rs** [comment]
Good point, you would still pay for one extra copy, but it's better than nothing. String interning might work for some very common header values, I guess it could save a couple alloc per request with common content types and accept header values for instance.

Before spending more time on this, I'd love to hear from @bradfitz about the `strings.Builder.Slice` idea and if not, what is the probability to having more complex implementation as suggested above accepted.

**Comment by bcmills** [comment]
@rs, `strings.Builder.Slice` seems too prone to misuse: much too easy to accidentally create multiple underlying strings (due to growth) when the user really only intended to retain slices of a single big string.

**Comment by rs** [comment]
Agreed but it would still be safe and you would still save allocations by creating those substring from a common buffer.

**Comment by kokizzu** [comment]
Possibly related, this allocation grows to 200MB over a week and cannot be reclaimed by GC:
![image](https://user-images.githubusercontent.com/1061610/106223575-49165380-6214-11eb-8cb2-c612a845a848.png)

**Comment by davecheney** [comment]
> Possibly related, this allocation grows to 200MB over a week and cannot be reclaimed by GC:

How are you asserting that this allocation cannot be garbage collected?

**Comment by kokizzu** [comment]
i run `github.com/google/gops` on the service:
```
$ gops pprof-heap 1
Profile dump saved to: /tmp/heap_profile363561508
$ gops gc 1 # force GC
$ gops pprof-heap 1
Profile dump saved to: /tmp/heap_profile269530696
```
copy both heap_profile and compare, it doesn't decrease '__')
![image](https://user-images.githubusercontent.com/1061610/106256135-7af7dc00-624d-11eb-87ce-f8f169e53b53.png)

**Comment by kokizzu** [comment]
my friend found it https://rover.rocks/golang-memory-leak '__') it was gin's context.Set that cache trashing that holds the memory

**Comment by s-tokutake** [comment]
We are facing the same problem as above. However, we does not use  gin-gonic/gin.

![image](https://user-images.githubusercontent.com/5404402/114252072-78003100-99de-11eb-94ba-eae9cc7abdd2.png)

**Comment by josharian** [comment]
Do you have any statistics about what the header values are? If there are a handful of very common values, perhaps we could hard-code some strings to re-use instead of allocate. (Like string interning, but hard-coded.)

**Comment by jonaz** [comment]
We are also seeing a majority of the memory being used in ReadMIMEHeader in a simple proxy application. I have yet to investigate if clients send unusually large headers (for example many cookies). 
![2022-03-25-075812_2024x1120_scrot](https://user-images.githubusercontent.com/1146389/160072570-8a669a35-c6da-48f6-bdcb-12bfaf14b0ec.png)

**Comment by CAFxX** [comment]
Spent a bit on this and, unfortunately, there does not seem to be a lot that can be done. The only approach I found that worked was interning, so this may be a good candidate for some form of string interning (not exposed to users):

```name                             old time/op    new time/op    delta
ReadMIMEHeader/client_headers-4    2.38µs ± 9%    2.26µs ±11%   -5.35%  (p=0.000 n=26+29)
ReadMIMEHeader/server_headers-4    2.10µs ± 8%    1.95µs ± 9%   -7.28%  (p=0.000 n=28+26)
Uncommon-4                          549ns ± 7%     522ns ±10%   -4.99%  (p=0.000 n=29+28)

name                             old alloc/op   new alloc/op   delta
ReadMIMEHeader/client_headers-4    1.53kB ± 0%    0.95kB ± 1%  -38.41%  (p=0.000 n=29+28)
ReadMIMEHeader/server_headers-4    1.10kB ± 0%    0.92kB ± 0%  -16.51%  (p=0.000 n=30+30)
Uncommon-4                           468B ± 0%      435B ± 0%   -7.05%  (p=0.000 n=30+27)

name                             old allocs/op  new allocs/op  delta
ReadMIMEHeader/client_headers-4      14.0 ± 0%       3.0 ± 0%  -78.57%  (p=0.000 n=30+30)
ReadMIMEHeader/server_headers-4      14.0 ± 0%       3.0 ± 0%  -78.57%  (p=0.000 n=30+30)
Uncommon-4                           5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=30+30)
```

I have pushed the poc [here](https://github.com/CAFxX/go/commit/e35283084e662b1e7ce57d9fd7b3c88a2cf2c845). If we think this may be useful I can try to refine it properly (the current interning logic is *really* crude) and send a CL.

**Comment by GitRowin** [comment]
Same issue here:
![profile001](https://user-images.githubusercontent.com/20298291/168087234-24bf9425-d2d5-4940-be0c-f4263dafe080.png)
Compiled using Go 1.18 if that matters.
